official/mnist/mnist.py

```diff
@@ -153,55 +165,122 @@ def train_input_fn():
-        FLAGS.batch_size).make_one_shot_iterator().get_next()
+    return dataset.test(flags.data_dir).batch(
+        flags.batch_size).make_one_shot_iterator().get_next()
 
   eval_results = mnist_classifier.evaluate(input_fn=eval_input_fn)
   print()
   print('Evaluation results:\n\t%s' % eval_results)
 
   # Export the model
-  if FLAGS.export_dir is not None:
+  if flags.export_dir is not None:
     image = tf.placeholder(tf.float32, [None, 28, 28])
     input_fn = tf.estimator.export.build_raw_serving_input_receiver_fn({
         'image': image,
     })
-    mnist_classifier.export_savedmodel(FLAGS.export_dir, input_fn)
+    mnist_classifier.export_savedmodel(flags.export_dir, input_fn)
 
 
-if __name__ == '__main__':
-  parser = argparse.ArgumentParser()
-  parser.add_argument(
-      '--batch_size',
-      type=int,
-      default=100,
-      help='Number of images to process in a batch')
-  parser.add_argument(
-      '--data_dir',
-      type=str,
-      default='/tmp/mnist_data',
-      help='Path to directory containing the MNIST dataset')
-  parser.add_argument(
-      '--model_dir',
-      type=str,
-      default='/tmp/mnist_model',
-      help='The directory where the model will be stored.')
-  parser.add_argument(
-      '--train_epochs', type=int, default=40, help='Number of epochs to train.')
-  parser.add_argument(
-      '--data_format',
-      type=str,
-      default=None,
-      choices=['channels_first', 'channels_last'],
-      help='A flag to override the data format used in the model. channels_first '
-      'provides a performance boost on GPU but is not always compatible '
-      'with CPU. If left unspecified, the data format will be chosen '
-      'automatically based on whether TensorFlow was built for CPU or GPU.')
-  parser.add_argument(
-      '--export_dir',
-      type=str,
-      help='The directory where the exported SavedModel will be stored.')
+################################################################################
+# Multi-GPU helpers
+################################################################################
+def model_fn_multi(features, labels, mode, params):
+  """Wrapper for the model_fn that sets the optimizer for multi-GPU.
+
+  There are two key steps: wrap the optimizer, then replicate the model
+  function.
+  """
+  return model_fn(features, labels, mode, params, get_optimizer_multi)
+
+
+def get_optimizer_multi(learning_rate):
+  optimizer = tf.train.AdamOptimizer(learning_rate)
+  return tf.contrib.estimator.TowerOptimizer(optimizer)
+
+
+def get_batch_size_multi(current_size):
+  """For multi-gpu, batch-size must be a multiple of the number of
+  available GPUs.
+
+  TODO(karmel): This should eventually be handled by replicate_model_fn
+  directly. For now, doing the work here.
+  """
+  remainder = current_size % _get_num_gpu()
+  return current_size - remainder
+
 
+def _get_num_gpu():
+  """TODO(karmel): This should eventually be handled by replicate_model_fn
```

Comment:
Good point. I feel like we should have some explanation of why this is required here, but I can do that as a comment rather than a TODO, which will be cleaner/less concerning?